bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -17,220 +17,45 @@
 //!   performing the handshake until they return success or failure. See
 //!   [`VerifyHostNameCallback`] as an example.
-//! * "async" callbacks return a [`Poll`] and should not block the task performing the handshake.
-//!   They will be polled until they return [`Poll::Ready`].
+//! * "async" callbacks return a [Poll](`core::task::Poll`) and should not block the task performing the handshake.
+//!   They will be polled until they return [Poll::Ready](`core::task::Poll::Ready`).
 //!   [Connection::waker()](`crate::connection::Connection::waker()`)
 //!   can be used to register the task for wakeup. See [`ClientHelloCallback`] as an example.
 
-use crate::{
-    config::Config,
-    connection::{Connection, InternalConnectionFuture},
-    enums::CallbackResult,
-    error::Error,
-};
-use core::{mem::ManuallyDrop, ptr::NonNull, task::Poll, time::Duration};
-use pin_project_lite::pin_project;
+use crate::{config::Context, connection::Connection};
+use core::{mem::ManuallyDrop, ptr::NonNull, time::Duration};
 use s2n_tls_sys::s2n_connection;
-use std::{future::Future, pin::Pin};
+
+mod async_cb;
+pub use async_cb::*;
+
+mod client_hello;
+pub use client_hello::*;
 
 /// Convert the connection pointer provided to a callback into a Connection
-/// useable with the Rust bindings.
+/// and Context useable with the Rust bindings.
 ///
 /// # Safety
 ///
 /// This must ONLY be used for connection pointers provided to callbacks,
 /// which can be assumed to point to valid Connections because the
 /// callbacks were configured through the Rust bindings.
-pub(crate) unsafe fn with_connection<F, T>(conn_ptr: *mut s2n_connection, action: F) -> T
+pub(crate) unsafe fn with_context<F, T>(conn_ptr: *mut s2n_connection, action: F) -> T
 where
-    F: FnOnce(&mut Connection) -> T,
+    F: FnOnce(&mut Connection, &mut Context) -> T,
 {
     let raw = NonNull::new(conn_ptr).expect("connection should not be null");
     let mut conn = Connection::from_raw(raw);
-    let r = action(&mut conn);
+    let mut config = conn.config().expect("config should not be null");
+    let context = config.context_mut();
+    let r = action(&mut conn, context);
     // Since this is a callback, it receives a pointer to the connection
     // but doesn't own that connection or control its lifecycle.
     // Do not drop / free the connection.
     let _ = ManuallyDrop::new(conn);
     r
 }
 
-/// Begins execution of an asyc callback.
-///
-/// Polls the async callback once, then registers it for later retries if
-/// necessary.
-///
-/// The C-style callback method passed to the underlying s2n-tls implementation
-/// should call this method instead of using the Rust callback implementation
-/// directly. The C-style callback will only execute once, so the underlying
-/// poll implementation should ensures that the Rust callback is polled until
-/// it completes.
-///
-/// Using [`config::set_client_hello_callback`] as an example, the execution
```

Comment:
Is any of this comment still relevant? Seems like we're losing a lot of documentation by deleting this.

Reply:
Mostly still relevant, but I also think unnecessarily verbose / complicated. I think the diagram below kind of got out of hand.

I tried to rewrite the comment here: https://github.com/aws/s2n-tls/blob/2a5e15299f8b9c20878e8e5025bfd1482ce4362b/bindings/rust/s2n-tls/src/callbacks/async_cb.rs#L6-L15

Reply:
But if you think an updated version of that diagram would be helpful, I can add it back.

Reply:
Actually I think the updated comment is clear, i just wasn't sure if we were losing anything since the old one is a lot longer.